File: Object-C/BlackFireIM/Example/ChatDetail/Controller/BFCallMessageCellData.m
//
//  MSCllMessageCellData.m
//  BlackFireIM
//
//  Created by benny wang on 2021/8/18.
//

#import "BFCallMessageCellData.h"
#import "MSIMSDK-UIKit.h"
#import <MSIMSDK/MSIMSDK.h>

@interface BFCallMessageCellData()

@property (nonatomic, assign) CGRect noticeFrame;

@property (nonatomic, assign) CGRect iconFrame;

@end

@implementation BFCallMessageCellData

- (instancetype)initWithDirection:(TMsgDirection)direction
{
    self = [super initWithDirection:direction];
    if (self) {
        
    }
    return self;
}

- (UIImage *)iconImage
{
    if (self.callType == MSCallType_Voice) {
        return self.direction == MsgDirectionIncoming ? [UIImage imageNamed:TUIKitResource(@"call_decline")] : [UIImage imageNamed:TUIKitResource(@"call_decline_white")];
    }else {
        return self.direction == MsgDirectionIncoming ? [UIImage imageNamed:TUIKitResource(@"video_right")] : [UIImage imageNamed:TUIKitResource(@"video_left")];
    }
}

- (CGSize)contentSize
{
    UIEdgeInsets contentInset = self.direction == MsgDirectionIncoming ? UIEdgeInsetsMake(10, 16, 10, 14) : UIEdgeInsetsMake(10, 14, 10, 16);
    CGSize size = [self.notice textSizeIn:CGSizeMake(TTextMessageCell_Text_Width_Max, MAXFLOAT) font:[UIFont systemFontOfSize:16]];
    if (self.direction == MsgDirectionIncoming) {
        self.iconFrame = CGRectMake(contentInset.left, contentInset.top, size.height, size.height);
        self.noticeFrame = CGRectMake(CGRectGetMaxX(self.iconFrame) + 5, contentInset.top, size.width, size.height);
    }else {
        self.noticeFrame = CGRectMake(contentInset.left, contentInset.top, size.width, size.height);
        self.iconFrame = CGRectMake(CGRectGetMaxX(self.noticeFrame) + 5, contentInset.top, size.height, size.height);
    }
    size.width += contentInset.left+contentInset.right + size.height + 5;
    size.height += contentInset.top+contentInset.bottom;
    
    return size;
}

- (NSString *)reuseId
{
    return @"TCallMessageCell";
}

@end

File: Swift/MSIM-Swift/ChatDetail/BFCallMessageCellData.swift
//
//  BFCallMessageCellData.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/10/3.
//

import UIKit

class BFCallMessageCellData: MSBubbleMessageCellData {

    var callType: MSCallType = .voice
    
    var notice: String = ""
    
    var iconImage: UIImage? {
        if callType == .voice {
            return direction == .inComing ? UIImage.bf_imageNamed(name: "call_decline") : UIImage.bf_imageNamed(name: "call_decline_white")
        }else {
            return direction == .inComing ? UIImage.bf_imageNamed(name: "video_right") : UIImage.bf_imageNamed(name: "video_left")
        }
    }
    
    private(set) var noticeFrame: CGRect = .zero
    
    private(set) var iconFrame: CGRect = .zero
    
    override init(direction: TMsgDirection) {
        super.init(direction: direction)
    }
    
    override func contentSize() -> CGSize {
        let contentInset = direction == .inComing ? UIEdgeInsets(top: 10, left: 16, bottom: 10, right: 14) : UIEdgeInsets(top: 10, left: 14, bottom: 10, right: 16)
        var size = (notice as NSString).textSize(in: CGSize(width: MSMcros.TTextMessageCell_Text_Width_Max, height: CGFloat.greatestFiniteMagnitude), font: .systemFont(ofSize: 16))
        if direction == .inComing {
            iconFrame = CGRect(x: contentInset.left, y: contentInset.top, width: size.height, height: size.height)
            noticeFrame = CGRect(x: iconFrame.maxX + 5, y: contentInset.top, width: size.width, height: size.height)
        }else {
            noticeFrame = CGRect(x: contentInset.left, y: contentInset.top, width: size.width, height: size.height)
            iconFrame = CGRect(x: noticeFrame.maxX + 5, y: contentInset.top, width: size.height, height: size.height)
        }
        size.width += contentInset.left + contentInset.right + size.height + 5
        size.height += contentInset.top + contentInset.bottom
        
        return size
    }
    
    override var reUseId: String {
        return "TCallMessageCell"
    }
}
